ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -442,9 +449,12 @@ private void maybeWriteToChannel() {
       // if there are chunks available to be written out, do now.
       if (firstChunk.isComplete() && readCalled) {
         while (operationException.get() == null && chunkIndexToBuf.containsKey(indexOfNextChunkToWriteOut)) {
-          ByteBuf byteBuf = chunkIndexToBuf.get(indexOfNextChunkToWriteOut);
-          asyncWritableChannel.write(byteBuf.nioBuffer(), chunkAsyncWriteCallback);
-          indexOfNextChunkToWriteOut++;
+          ByteBuf byteBuf = chunkIndexToBuf.remove(indexOfNextChunkToWriteOut);
+          if (byteBuf != null) {
+            asyncWritableChannel.write(byteBuf.nioBuffer(), chunkAsyncWriteCallback);
+            chunkIndexToBufWaitingForRelease.put(indexOfNextChunkToWriteOut, byteBuf);
```

Comment:
should this call come before `asyncWritableChannel.write` in case write calls `chunkAsyncWriteCallback` before the buff is added to the map?

Reply:
good catch.